man/richacl.7
.\"
.\" RichACL Manual Pages
.\"
.\" Copyright (C) 2015,2016  Red Hat, Inc.
.\" Written by Andreas Gruenbacher <agruenba@redhat.com>
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual.  If not, see
.\" <http://www.gnu.org/licenses/>.
.\"
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.TH RICHACL 7 2016-02-23 "Linux" "Rich Access Control Lists"
.SH NAME
richacl \- Rich Access Control Lists
.SH DESCRIPTION
Rich Access Control Lists (RichACLs) are an extension of the traditional POSIX
file permission model to support
.URL https://tools.ietf.org/rfc/rfc5661.txt "NFSv4 Access Control Lists"
on local and remote-mounted filesystems.
.PP
A RichACL can always be translated into an equivalent NFSv4 ACL which grants
the same permissions.
.PP
RichACLs can be enabled on supported local filesystems (currently, ext4 and
XFS).  The in-kernel NFSv4 server and client also include RichACL support.
.PP
Enabling RichACLs disables POSIX Access Control Lists; the two ACL models
cannot coexist on the same filesystem.
.PP
When used on a filesystem that does not support RichACLs, the
.BR getrichacl (1)
and
.BR setrichacl (1)
utilities will operate on the file permission bits instead:
.BR getrichacl (1)
will display the file permission bits as a RichACL; when a RichACL
is set with
.BR setrichacl (1)
which can be represented exactly by the file permission bits,
.BR setrichacl (1)
will set the file permission bits instead.  An attempt to set a RichACL that
cannot be represented exactly by the file permission bits results in an error.
.PP
This man-page describes the structure, concepts and algorithms involved in RichACLs. For examples, please refer to
.BR richaclex (7).
.SS Structure of RichACLs
RichACLs consist of a number of ACL entries, three file masks, and a set of
flags specifying attributes of the ACL as a whole (by contrast with the
per-ACL-entry flags described below).
.PP
Each of the ACL entries allows or denies some permissions to a particular user,
group, or special identifier. An ACL entry consists of four fields:
.IP \(bu 2
A tag which specifies the user, group, or special identifier the entry applies
to.  Special identifiers can be the file owner
.RB ( owner@ ),
the owning group
.RB ( group@ ),
or everyone
.RB ( everyone@ ).
.IP \(bu
A set of permissions the entry allows or denies.
.IP \(bu
A set of flags that indicate whether the user or group identifier is mapped or
unmapped, and whether the entry has been and can be inherited.
.IP \(bu
A type field indicating whether the entry allows or denies the permissions
specified in the second field.
.PP
The owner, group, and other file masks further control which permissions the
ACL grants, subject to the
.BR masked "\ (" m )
and
.BR write_through "\ (" w )
ACL flags: when the permissions of a file or directory are changed with
.BR chmod (2),
the file masks are set based on the new file mode, and the
.B masked
and
.B write_through
ACL flags are set. Likewise, when a new file or directory inherits an ACL from
its parent directory, the file masks are set to the intersection between the
permissions granted by the inherited ACL and the
.I mode
parameter as given to
.BR open "(2), " mkdir (2),
and similar, and the
.B masked
ACL flag is set. In both cases, the file masks limit the permissions that the
ACL will grant.
.PP
The purpose of the file masks is to allow traditional POSIX applications which
are unaware of RichACLs to place limits on the permissions granted by the
RichACL without causing the ACL entries to be modified.  Without the file mask,
the only alternative would be for the kernel to directly modify the ACL
entries.  However, this latter approach could have a \(lqhysteresis effect\(rq,
whereby a RichACL-unaware application that performs a transition from one file
mode to another file mode and then back to the original file mode could cause
destruction of the permission information in ACL entries.  When creating files
or directories, the
.I mode
parameter to
.BR open "(2), " mkdir (2),
and similar would have the same effect.
.PP
Note that entries with the identifier
.B everyone@
apply to all processes, whereas the \(lqother\(rq file permissions and
\(lqother\(rq (\fBACL_OTHER\fP) entries in POSIX ACLs apply to all processes
which are not the owner, are not in the owning group, and do not match a user
or group mentioned in the ACL.
.PP
Unlike POSIX ACLs, RichACLs do not have separate \(lqaccess\(rq ACLs that
define the access permissions and \(lqdefault\(rq ACLs that define the
inheritable permissions.  Instead, flags on each ACL entry determine whether
the entry is effective during access checks and/or inheritable.
.SS ACL flags
The following ACL flags are defined:
.TP
.BR masked "\ (" m )
When set, the file masks define upper limits on the permissions the ACL may
grant.  When not set, the file masks are ignored.
.TP
.BR write_through "\ (" w )
When this flag and the
.B masked
flag are both set, the owner and other file masks define the actual permissions
granted to the file owner and to others instead of defining an upper limit.
When the
.B masked
flag is not set, the
.B write_through
flag has no effect.
.TP
.BR auto_inherit "\ (" a )
Automatic Inheritance is enabled.  See
.IR "Automatic Inheritance" .
.TP
.BR protected "\ (" p )
The ACL is protected from modification by Automatic
Inheritance.
.TP
.BR defaulted "\ (" d )
Indicates that the ACL has been assigned in an application-dependent way when
the file or directory was created; it has neither been inherited nor set
explicitly. ACLs of files created on Linux will never have the
.B defaulted
flag set, but the flag may be set on files created on or copied over from
other operating systems.  When this flag is set for an ACL, Automatic
Inheritance will completely replace the ACL.
.SS ACL entry flags

The following flags on ACL entries are defined:
.TP
.BR file_inherit "\ (" f )
When this flag appears in the ACL entry of a directory, then:
.RS
.IP \(bu 2
That entry is inherited by new files created in the directory such that the
permissions of the entry apply to the file (the inheritance flags of the
inherited entry are cleared).
.IP \(bu
The entry is is inherited by new subdirectories created in the directory such
that the permissions of the entry will apply to new files created in the
subdirectory.
.RE
.TP
.BR dir_inherit "\ (" d )
When this flag appears in the ACL entry of a directory, then that entry is
inherited by new subdirectories created in the directory such that the
permissions of the entry apply to the subdirectory (the
.B inherit_only
flag is cleared).
.TP
.BR no_propagate "\ (" n )
Inheritance stops at the next subdirectory level.  When a file or directory
inherits an entry that has this flag set, the
.BR file_inherit ", " dir_inherit ", " no_propagate ", and " inherit_only
flags are cleared.
.TP
.BR inherit_only "\ (" i )
The entry defines inheritable permissions only and is ignored for access
checking.  When a file or directory inherits an entry that has this flag set,
the flag is cleared.
.TP
.BR inherited "\ (" a )
The entry has been automatically inherited from the parent directory.
.TP
.BR unmapped "\ (" u )
The user or group identifier is a textual string and is not mapped to a numeric
user or group identifier.  ACLs with unmapped identifiers can occur on NFSv4
mounted filesystems when the client cannot determine numeric user or group
identifiers for some of the NFSv4 user@domain or group@domain who values.  They
cannot be assigned to local files or directories.
.SS Permissions
The following permissions are defined for RichACL entries and for the three
file masks:
.TP
.BR read_data " / " list_directory "\ (" r )
For a file: read the data of the file.
For a directory: list the contents of the directory.
.TP
.BR write_data " / " add_file "\ (" w )
For a file: modify the data of the file; does not include opening the file in
append mode.
For a directory: add a new file in the directory.
.TP
.BR append_data " / " add_subdirectory "\ (" p )
For a file: open the file in append mode.
For a directory: create a subdirectory in the directory.
.TP
.BR execute "\ (" x )
For a file: execute the file.
For a directory: traverse (search) the directory.
.TP
.BR delete_child "\ (" d )
Delete a file or directory within a directory.  This permission is meaningful
only for directories.
.TP
.BR delete "\ (" D )
Delete the file or directory.
.TP
.BR read_attributes "\ (" a )
Read basic attributes of a file or directory (see
.BR stat (2)).
This permission is defined by NFSv4.  It is stored, but ignored.  Reading basic
attributes of files and directories is always allowed on Linux.
.TP
.BR write_attributes "\ (" A )
Change the times associated with a file or directory to an arbitrary value.
This permission is always implicitly granted to the file owner.
.TP
.BR read_acl "\ (" c )
Read the ACL of a file or directory. Like reading the basic file attributes (the
.B read_attributes
permission), reading ACLs is always allowed on Linux.
.TP
.BR write_acl "\ (" C )
Change the ACL or file mode of a file or directory.
.TP
.BR write_owner "\ (" o )
Take ownership of a file or directory.  Change the owning group of a file or
directory to a group of which the calling process is a member.
.TP
.ad l
.BR read_named_attrs "\ (" R "), " write_named_attrs "\ (" W "), " synchronize "\ (" S "), " write_retention "\ (" e "), " write_retention_hold "\ (" E )
.ad
These permissions are defined by NFSv4 and NFSv4.1.  They are stored, but ignored.
.PP
For the
.BR r ", " w ", and " p
permissions, which have different long forms for files and directories, the
.BR getrichacl (1)
utility will output the appropriate form(s) depending on the context.
The
.BR setrichacl (1)
utility will accept either form for any file type.
.SS Text form
The common textual representation of a RichACL consists of the colon-separated
fields of the ACL flags, file masks, and ACL entries in the following
format:
.TP
\fBflags:\fR\fIacl_flags\fR
The ACL flags.
.TP
\fBowner:\fR\fIperm\fR\fB::mask\fR, \fBgroup:\fR\fIperm\fR\fB::mask\fR, \fBother:\fR\fIperm\fR\fB::mask\fR
The file masks and their permissions.
.TP
\fIwho\fR\fB:\fR\fIperm\fR\fB:\fR\fIflags\fR\fB:allow\fR, \fIwho\fR\fB:\fR\fIperm\fR\fB:\fR\fIflags\fR\fB:deny\fR
For each ACL entry, who the entry applies to, the permissions of the entry, the
entry flags, and the entry type. The \fIwho\fR field is one of the following:
.RS
.IP \(bu 2
One of the special identifiers:
.BR owner@ ", " group@ ", or " everyone@ ,
.IP \(bu
A
.BR user: " or " u:
prefix followed by a user name or user ID that designates a specific user,
.IP \(bu
A
.BR group: " or " g:
prefix followed by a group name or group ID that designates a specific group.
.RE
.PP
The ACL flags, file masks, and ACL entries are comma, whitespace, or newline
separated.
.PP
Flags and permissions have single-letter as well as long forms, as listed under
.IR "ACL flags" ,
.IR "ACL entry flags" ,
and

doc/CHANGES
1.0.0 (14 Oct 2008)
* Initial version of nfs4acl from Andreas Gruenbacher and sgi.


richacl/richacl.c
/*
  Copyright (C) 2006, 2008, 2009, 2010  Novell, Inc.
  Written by Andreas Gruenbacher <agruen@suse.de>

  This program is free software; you can redistribute it and/or modify it
  under the terms of the GNU General Public License as published by the
  Free Software Foundation; either version 2, or (at your option) any
  later version.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License along
  with this library; if not, write to the Free Software Foundation, Inc.,
  59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
*/

/*
 * FIXME:
 * Make ls show a `+' for richacls (in coreutils).
 * Add a way to show only expicitly set acls, and hide inherited ones.
 * Convert a non-Automatic-Inheritance tree into an Automatic Inheritance one?
 */

#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <dirent.h>
#include <unistd.h>
#include <sys/xattr.h>
#include <ctype.h>
#include <pwd.h>
#include <grp.h>

#include "richacl.h"
#include "string_buffer.h"

static const char *progname;
int opt_repropagate;

void printf_stderr(const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);
}

#define richacl_for_each_entry_continue(_ace, _acl) \
	for ((_ace)++; \
	     (_ace) != (_acl)->a_entries + (_acl)->a_count; \
	     (_ace)++)

static inline int richace_is_inherited(const struct richace *ace)
{
	return ace->e_flags & RICHACE_INHERITED_ACE;
}

static void add_implicitly_granted_permissions(struct richacl *acl)
{
	struct richace *ace;

	richacl_for_each_entry(ace, acl) {
		unsigned int mask;

		mask = RICHACE_POSIX_ALWAYS_ALLOWED;
		if (richace_is_owner(ace))
			mask |= RICHACE_POSIX_OWNER_ALLOWED;

		if (richace_is_allow(ace))
			ace->e_mask |= mask;
		else if (richace_is_deny(ace))
			ace->e_mask &= ~mask;
	}
}

static void compute_masks(struct richacl *acl, int acl_has)
{
	unsigned int owner_mask = acl->a_owner_mask;
	unsigned int group_mask = acl->a_group_mask;
	unsigned int other_mask = acl->a_other_mask;

	if ((acl_has & RICHACL_TEXT_OWNER_MASK) &&
	    (acl_has & RICHACL_TEXT_GROUP_MASK) &&
	    (acl_has & RICHACL_TEXT_OTHER_MASK) &&
	    (acl_has & RICHACL_TEXT_FLAGS))
		return;

	if (!(acl_has & RICHACL_TEXT_FLAGS))
		acl->a_flags &= ~RICHACL_MASKED;
	richacl_compute_max_masks(acl);
	if (acl_has & RICHACL_TEXT_OWNER_MASK) {
		if (!(acl_has & RICHACL_TEXT_FLAGS) &&
		    (acl->a_owner_mask & ~owner_mask))
			acl->a_flags |= RICHACL_MASKED;
		acl->a_owner_mask = owner_mask;
	}
	if (acl_has & RICHACL_TEXT_GROUP_MASK) {
		if (!(acl_has & RICHACL_TEXT_FLAGS) &&
		    (acl->a_group_mask & ~group_mask))
			acl->a_flags |= RICHACL_MASKED;
		acl->a_group_mask = group_mask;
	}
	if (acl_has & RICHACL_TEXT_OTHER_MASK) {
		if (!(acl_has & RICHACL_TEXT_FLAGS) &&
		    (acl->a_other_mask & ~other_mask))
			acl->a_flags |= RICHACL_MASKED;
		acl->a_other_mask = other_mask;
	}
}

int modify_richacl(struct richacl **acl2, struct richacl *acl, int acl_has)
{
	struct richace *ace2, *ace;

	if (richacl_apply_masks(acl2))
		return -1;

	richacl_for_each_entry(ace, acl) {
		struct richacl *acl3;
		struct richace *ace3;

		richacl_for_each_entry(ace2, *acl2) {
			if (ace2->e_type == ace->e_type &&
			    richace_is_inherited(ace2) == richace_is_inherited(ace) &&
			    richace_is_same_identifier(ace, ace2)) {
				ace2->e_mask = ace->e_mask;
				ace2->e_flags = ace->e_flags;
				goto next_change;
			}
		}

		acl3 = richacl_alloc((*acl2)->a_count + 1);
		if (!acl3)
			return -1;
		acl3->a_flags = (*acl2)->a_flags;
		ace3 = acl3->a_entries;
		if (!(ace->e_flags & RICHACE_INHERITED_ACE)) {
			if (richace_is_deny(ace)) {
				/*
				 * Insert the new deny entry after the existing
				 * initial non-inherited deny entries.
				 */
				richacl_for_each_entry(ace2, *acl2) {
					if (!richace_is_deny(ace2) ||
					    richace_is_inherited(ace2))
						break;
					richace_copy(ace3++, ace2);
				}
			} else {
				/*
				 * Append the new allow entry at the end of the
				 * non-inherited aces.
				 */
				richacl_for_each_entry(ace2, *acl2) {
					if (richace_is_inherited(ace2))
						break;
					richace_copy(ace3++, ace2);
				}
			}
			richace_copy(ace3++, ace);
			ace2--;
			richacl_for_each_entry_continue(ace2, *acl2)
				richace_copy(ace3++, ace2);
		} else {
			struct richace *last_inherited;

			last_inherited = (*acl2)->a_entries + (*acl2)->a_count;
			while (last_inherited > (*acl2)->a_entries &&
			       richace_is_inherited(last_inherited - 1))
				last_inherited--;

			richacl_for_each_entry(ace2, *acl2) {
				if (ace2 == last_inherited)
					break;
				richace_copy(ace3++, ace2);
			}
			if (richace_is_deny(ace)) {
				/*
				 * Insert the new deny entry after the existing
				 * initial inherited deny entries.
				 */
				ace2--;
				richacl_for_each_entry_continue(ace2, *acl2) {
					if (!richace_is_deny(ace2))
						break;
					richace_copy(ace3++, ace2);
				}
			} else {
				/*
				 * Append the new allow entry at the end of the
				 * inherited aces.
				 */
				ace2--;
				richacl_for_each_entry_continue(ace2, *acl2)
					richace_copy(ace3++, ace2);
			}
			richace_copy(ace3++, ace);
			ace2--;
			richacl_for_each_entry_continue(ace2, *acl2)
				richace_copy(ace3++, ace2);
		}

		richacl_free(*acl2);
		*acl2 = acl3;

	next_change:
		/* gcc is unhappy without a statement behind the label ... */ ;
	}

	if (acl_has & RICHACL_TEXT_FLAGS)
		(*acl2)->a_flags = acl->a_flags;
	if (acl_has & RICHACL_TEXT_OWNER_MASK)
		(*acl2)->a_owner_mask = acl->a_owner_mask;
	if (acl_has & RICHACL_TEXT_GROUP_MASK)
		(*acl2)->a_group_mask = acl->a_group_mask;
	if (acl_has & RICHACL_TEXT_OTHER_MASK)
		(*acl2)->a_other_mask = acl->a_other_mask;
	compute_masks(*acl2, acl_has);

	return 0;
}

static int auto_inherit(const char *dirname, struct richacl *dir_acl)
{
	DIR *dir;
	struct richacl *dir_inheritable, *file_inheritable;
	struct dirent *dirent;
	char *path = NULL;
	size_t dirname_len;
	int status = 0;

	dir = opendir(dirname);
	if (!dir) {
		if (errno == ENOTDIR)
			return 0;
		return -1;
	}

	dirname_len = strlen(dirname);
	path = malloc(dirname_len + 2);
	if (!path)
		goto fail;
	sprintf(path, "%s/", dirname);

	errno = 0;
	file_inheritable = richacl_inherit(dir_acl, 0);
	if (!file_inheritable && errno != 0)
		goto fail;
	dir_inheritable = richacl_inherit(dir_acl, 1);
	if (!dir_inheritable && errno != 0)
		goto fail;

	while ((errno = 0, dirent = readdir(dir))) {
		struct richacl *old_acl = NULL, *new_acl = NULL;
		int isdir;
		char *p;

		if (!strcmp(dirent->d_name, ".") ||
		    !strcmp(dirent->d_name, ".."))
			continue;

		p = realloc(path, strlen(dirname) + strlen(dirent->d_name) + 2);
		if (!p)
			goto fail;
		path = p;
		strcpy(path + dirname_len + 1, dirent->d_name);

		if (dirent->d_type == DT_UNKNOWN) {
			struct stat st;

			if (lstat(path, &st))
				goto fail2;
			dirent->d_type = IFTODT(st.st_mode);
		}
		if (dirent->d_type == DT_LNK)
			continue;
		isdir = (dirent->d_type == DT_DIR);

		old_acl = richacl_get_file(path);
		if (!old_acl) {
			if (errno == ENODATA || errno == ENOTSUP || errno == ENOSYS)
				goto next;
			goto fail2;
		}
		if (!richacl_is_auto_inherit(old_acl))
			goto next;
		if (old_acl->a_flags & RICHACL_PROTECTED) {
			if (!opt_repropagate)
				goto next;
			new_acl = old_acl;
			old_acl = NULL;
		} else {
			int equal;
			if (old_acl->a_flags & RICHACL_DEFAULTED) {
				/* RFC 5661: An application performing
				 * automatic inheritance takes the
				 * RICHACL_DEFAULTED flag as a sign that the ACL
				 * should be completely replaced by one
				 * generated using the automatic inheritance
				 * rules. */

				free(old_acl);
				old_acl = richacl_alloc(0);
				old_acl->a_flags |= RICHACL_AUTO_INHERIT;
			}
			new_acl = richacl_auto_inherit(old_acl,
					isdir ? dir_inheritable :
						file_inheritable);
			equal = !richacl_compare(old_acl, new_acl);
			if (equal && !opt_repropagate)
				goto next;
			if (!equal && richacl_set_file(path, new_acl))
				goto fail2;
		}

		if (isdir)
			if (auto_inherit(path, new_acl))
				goto fail2;

	next:
		free(old_acl);
		free(new_acl);
		continue;

	fail2:
		perror(path);
		free(old_acl);
		free(new_acl);
		free(path);
		status = -1;
	}
	if (errno != 0) {
		perror(dirname);
		status = -1;
	}
	free(path);
	closedir(dir);
	return status;

fail:
	perror(basename(progname));
	free(path);
	closedir(dir);
	return -1;
}

static struct richacl *get_richacl(const char *file, mode_t mode)
{
	struct richacl *acl;

	acl = richacl_get_file(file);
	if (!acl) {
		if (errno == ENOTSUP &&
		    (getxattr(file, "system.posix_acl_access", NULL, 0) >= 0 ||
		    (S_ISDIR(mode) &&
		     getxattr(file, "system.posix_acl_default", NULL, 0) >= 0))) {
			fprintf(stderr, "%s: POSIX ACL(s) exist\n", file);
			errno = 0;
			return NULL;
		} else if (errno == ENODATA || errno == ENOTSUP || errno == ENOSYS)
			acl = richacl_from_mode(mode);
	}
	return acl;
}

static int set_richacl(const char *path, struct richacl *acl)
{
	if (richacl_set_file(path, acl)) {
		struct stat st;

		if (stat(path, &st))
			return -1;
		if (!richacl_equiv_mode(acl, &st.st_mode))
			return chmod(path, st.st_mode);
		/* FIXME: We could try POSIX ACLs here as well. */
		return -1;
	}
	if (richacl_is_auto_inherit(acl)) {
		int ret;

		ret =  auto_inherit(path, acl);
		return ret;
	}
	return 0;
}

int format_for_mode(mode_t mode)
{
	if (S_ISDIR(mode))
		return RICHACL_TEXT_DIRECTORY_CONTEXT;
	else
		return RICHACL_TEXT_FILE_CONTEXT;
}

static int print_richacl(const char *file, struct richacl **acl,
			 struct stat *st, int fmt)
{
	char *text;

	if (!(fmt & RICHACL_TEXT_SHOW_MASKS)) {
		if (richacl_apply_masks(acl))
			goto fail;
	}
	text = richacl_to_text(*acl, fmt | format_for_mode(st->st_mode));
	if (!text)
		goto fail;
	printf("%s:\n", file);
	puts(text);
	free(text);
	return 0;

fail:
	return -1;
}

void remove_filename(struct string_buffer *buffer)
{
	char *c, *end;

	for (c = buffer->buffer, end = buffer->buffer + buffer->offset;
	     c != end;
	     c++) {
		if (*c == '\n')
			break;
		if (*c == ':') {
			c++;
			if (c == end)
				break;
			if (*c != '\n')
				continue;
			c++;
			memmove(buffer->buffer, c, end - c + 1);
			buffer->offset = end - c;
		}
	}
}

static struct option long_options[] = {
	{"access",		2, 0, 'a'},
	{"get",			0, 0, 'g'},
	{"modify",		1, 0, 'm'},
	{"modify-file",		1, 0, 'M'},
	{"set",			1, 0, 's'},
	{"set-file",		1, 0, 'S'},
	{"remove",		0, 0, 'r'},
	{"long",		0, 0, 'l'},
	{"raw",			0, 0,  1 },
	{"dry-run",		0, 0,  2 },
	{"full",                0, 0,  3 },
	{"unaligned",		0, 0,  4 },
	{"numeric-ids",		0, 0,  5 },
	{"version",		0, 0, 'v'},
	{"help",		0, 0, 'h'},
	{ NULL,			0, 0,  0 }
};

static void synopsis(int help)
{
	FILE *file = help ? stdout : stderr;

	fprintf(file, "SYNOPSIS: %s [options] {command} file ...\n",
		basename(progname));
	if (!help) {
		fprintf(file, "Try `%s --help' for more information.\n",
			basename(progname));
		exit(1);
	}
	fprintf(file,
"\n"
"Commands:\n"
"  --get, -g   Display the ACL of file(s). Multiple ACL entries are separated\n"
"              by newline.\n"
"  --modify acl_entries, -m acl_entries\n"
"              Modify the ACL of file(s) by replacing existing entries with\n"
"              the entries in acl_entries, and adding all new entries. When\n"
"              the permissions of an entry are empty, remove the entry.\n"
"  --set acl, -s acl\n"
"              Set the ACL of file(s) to acl. Multiple ACL entries are separated\n"
"              by whitespace or commas.\n"
"  --modify-file acl_entries_file, -M acl_entries_file\n"
"  --set-file acl_file, -S acl_file\n"
"              Identical to --modify / --set, but read the ACL from a file\n"
"              instead. If the file is `-', read from standard input.\n"
"  --remove, -r\n"
"              Remove the ACL of file(s).\n"
"  --access[=user[:group:...]}, -a[user[:group:...]}\n"
"              Show which permissions the caller or a specified user has for\n"
"              file(s).  When a list of groups is given, this overrides the\n"
"              groups the user is in.\n"
"              capabilities. \n"
"  --version, -v\n"
"              Display the version of %s and exit.\n"
"  --help, -h  This help text.\n"
"\n"
"Options:\n"
"  --long, -l  Display access masks and flags in their long form.\n"
"  --full      Also show permissions which are always implicitly allowed.\n"
"              Always allow those permissions when setting/modifying acls.\n"
"  --raw       Show acls as stored on the file system including the file masks.\n"
"              Implies --full.\n"
"  --unaligned\n"
"              Do not align acl entries or pad missing permissions with '-'.\n"
"  --numeric-ids\n"
"              Display numeric user and group IDs instead of names.\n"
"\n"
"ACL entries are represented by colon separated <who>:<mask>:<flags>:<type>\n"
"fields. The <who> field may be \"owner@\", \"group@\", \"everyone@\", a user\n"
"name or ID, or a group name or ID. Groups have the identifier_group(g) flag\n"
"set in the <flags> field. The <type> field may be \"allow\" or \"deny\".\n"
"The <mask> and <flags> fields are lists of single-letter abbreviations or\n"
"slash-separated names, or a combination of both.\n"
"\n"
"ACL entry <mask> values are:\n"
"\tread_data (r), list_directory (r), write_data (w), add_file (w),\n"
"\texecute (x), append_data (p), add_subdirectory (p), delete_child (d),\n"
"\tdelete (D), read_attributes (a), write_attributes (A), read_xattr (R),\n"
"\twrite_xattr (W), read_acl (c), write_acl (C), write_owner(o),\n"
"\tsynchronize (S), write_retention (e), write_retention_hold (E)\n"
"\n"
"ACL entry <flags> values are:\n"
"\tfile_inherit (f), dir_inherit (d),\n"
"\tno_propagate (n), inherit_only (i),\n"
"\tidentifier_group (g), inherited (a)\n"
"\n"
"ACL flag values are:\n"
"\tmasked (m), auto_inherit (a), protected (p), defaulted (d)" /* ", posix_mapped (P)" */ "\n",
	basename(progname));
	exit(0);
}

int main(int argc, char *argv[])
{
	int opt_get = 0, opt_remove = 0, opt_access = 0, opt_dry_run = 0;
	int opt_modify = 0, opt_set = 0;
	int opt_full = 0;
	char *opt_user = NULL;
	char *acl_text = NULL, *acl_file = NULL;
	int format = RICHACL_TEXT_SIMPLIFY | RICHACL_TEXT_ALIGN;
	uid_t user = -1;
	gid_t *groups = NULL;
	int n_groups = -1;
	int status = 0;
	int c;

	struct richacl *acl = NULL;
	int acl_has;

	progname = argv[0];

	while ((c = getopt_long(argc, argv, "gm:M:s:S:a::rlvh",
				long_options, NULL)) != -1) {
		switch(c) {
			case 'g':
				opt_get = 1;
				break;
			case 'm':
				opt_modify = 1;
				acl_text = optarg;
				break;

			case 'M':
				opt_modify = 1;
				acl_file = optarg;
				break;
			case 's':
				opt_set = 1;
				acl_text = optarg;
				break;

			case 'S':
				opt_set = 1;
				acl_file = optarg;
				break;

			case 'r':
				opt_remove = 1;
				break;

			case 'a':  /* --access */
				opt_access = 1;
				opt_user = optarg;
				break;

			case 'l':
				format |= RICHACL_TEXT_LONG;
				break;

			case 'v':
				printf("%s %s\n", basename(progname), VERSION);
				exit(0);

			case 'h':
				synopsis(1);
				break;

			case 1:  /* --raw */
				format |= RICHACL_TEXT_SHOW_MASKS;
				format &= ~RICHACL_TEXT_SIMPLIFY;
				break;

			case 2:  /* --dry-run */
				opt_dry_run = 1;
				break;

			case 3:  /* --full */
				format &= ~RICHACL_TEXT_SIMPLIFY;
				opt_full = 1;
				break;

			case 4:  /* --unaligned */
				format &= ~RICHACL_TEXT_ALIGN;
				break;

			case 5:  /* --numeric-ids */
				format |= RICHACL_TEXT_NUMERIC_IDS;
				break;

			default:
				synopsis(0);
				break;
		}
	}
	if (opt_get + opt_remove + opt_modify + opt_set + opt_access != 1 ||
	    (acl_text ? 1 : 0) + (acl_file ? 1 : 0) > 1 ||
	    optind == argc)
		synopsis(optind != argc);

	if (acl_text) {
		acl = richacl_from_text(acl_text, &acl_has, printf_stderr);
		if (!acl)
			return 1;
		if (!opt_full)
			add_implicitly_granted_permissions(acl);
	}

	if (acl_file) {
		struct string_buffer *buffer;
		char buf[4097];
		FILE *file = stdin;
		ssize_t sz;

		buffer = alloc_string_buffer(1024);
		if (!buffer) {
			perror("");
			return 1;
		}

		if (strcmp(acl_file, "-")) {
			file = fopen(acl_file, "r");
			if (!file) {
				perror(acl_file);
				return 1;
			}
		}
		do {
			sz = fread(buf, 1, sizeof(buf) - 1, file);
			buf[sz] = 0;
			buffer_sprintf(buffer, "%s", buf);
		} while (!(feof(file) || ferror(file)));
		fclose(file);
		if (ferror(file)) {
			perror(acl_file);
			return 1;
		}

		remove_filename(buffer);
		acl = richacl_from_text(buffer->buffer, &acl_has, printf_stderr);
		if (!acl)
			return 1;
		free_string_buffer(buffer);
	}

	/* Compute all masks which haven't been set explicitly. */
	/* FIXME: how about --modify? */
	if (opt_set && acl)
		compute_masks(acl, acl_has);

	if (opt_user) {
		int n_groups_alloc;
		char *opt_groups;
		struct passwd *passwd = NULL;
		char *endp;

		opt_groups = strchr(opt_user, ':');
		if (opt_groups)
			*opt_groups++ = 0;

		user = strtoul(opt_user, &endp, 10);
		if (*endp) {
			passwd = getpwnam(opt_user);
			if (passwd == NULL) {
				fprintf(stderr, "%s: No such user\n", opt_user);
				exit(1);
			}
			user = passwd->pw_uid;
		} else
			user = -1;

		n_groups_alloc = 32;
		groups = malloc(sizeof(gid_t) * n_groups_alloc);
		if (!groups)
			goto fail;
		if (opt_groups) {
			char *tok;
			n_groups = 0;
			tok = strtok(opt_groups, ":");
			while (tok) {
				struct group *group;

				if (n_groups == n_groups_alloc) {
					gid_t *new_groups;
					n_groups_alloc *= 2;
					new_groups = realloc(groups, sizeof(gid_t) * n_groups_alloc);
					if (!new_groups)
						goto fail;
				}

				groups[n_groups] = strtoul(tok, &endp, 10);
				if (*endp) {
					group = getgrnam(tok);
					if (!group) {
						fprintf(stderr, "%s: No such group\n", tok);
						exit(1);
					}
					groups[n_groups] = group->gr_gid;
				}
				n_groups++;

				tok = strtok(NULL, ":");
			}
		} else {
			if (!passwd)
				passwd = getpwuid(user);
			if (passwd) {
				n_groups = n_groups_alloc;
				if (getgrouplist(passwd->pw_name, passwd->pw_gid,
					         groups, &n_groups) < 0) {
					free(groups);
					groups = malloc(sizeof(gid_t) * n_groups);
					if (getgrouplist(passwd->pw_name, passwd->pw_gid,
							 groups, &n_groups) < 0)
						goto fail;
				}
			} else
				n_groups = 0;
		}
	} else
		user = geteuid();

	for (; optind < argc; optind++) {
		const char *file = argv[optind];
		struct richacl *acl2 = NULL;
		struct stat st;

		if (opt_get || opt_modify || opt_access || opt_dry_run) {
			if (stat(file, &st))
				goto fail2;
		} else
			memset(&st, 0, sizeof(st));

		if (opt_set) {
			if (opt_dry_run) {
				if (print_richacl(file, &acl, &st, format))
					goto fail2;
			} else {
				if (set_richacl(file, acl))
					goto fail2;
			}
		} else if (opt_modify) {
			acl2 = get_richacl(file, st.st_mode);
			if (!acl2) {
				if (!errno)
					goto fail3;
				goto fail2;
			}
			if (modify_richacl(&acl2, acl, acl_has))
				goto fail2;
			if (opt_dry_run) {
				if (print_richacl(file, &acl2, &st, format))
					goto fail2;
			} else {
				if (set_richacl(file, acl2))
					goto fail2;
			}
		} else if (opt_remove) {
			if (removexattr(file, "system.richacl")) {
				if (errno != ENODATA)
					goto fail2;
			}
		} else if (opt_access) {
			unsigned int mask;
			char *mask_text;

			mask = richacl_access(file, &st, user, groups, n_groups);
			if (mask < 0)
				goto fail2;

			mask_text = richacl_mask_to_text(mask,
					format | format_for_mode(st.st_mode));
			printf("%s  %s\n", mask_text, file);
			free(mask_text);
		} else /* opt_get */ {
			acl2 = get_richacl(file, st.st_mode);
			if (!acl2) {
				if (!errno)
					goto fail3;
				goto fail2;
			}
			if (print_richacl(file, &acl2, &st, format))
				goto fail2;
		}
		richacl_free(acl2);
		continue;

	fail2:
		richacl_free(acl2);
		perror(file);
	fail3:
		status = 1;
	}

	richacl_free(acl);
	return status;

fail:
	perror(basename(progname));
	return 1;
}
